src/components/VerticalNavigation/constants.js

```diff
@@ -0,0 +1,105 @@
+const navContextTypes = {
+  idPath: PropTypes.string, // **
+  depth: PropTypes.oneOf(['primary', 'secondary', 'tertiary']), // *
+  primaryItem: PropTypes.shape(itemObjectTypes), // *
+  secondaryItem: PropTypes.shape(itemObjectTypes), // *
+  hiddenIcons: PropTypes.bool,
+  pinnableMenus: PropTypes.bool,
+  isMobile: PropTypes.bool,
+  selectedMobileDepth: PropTypes.oneOf([null, 'primary', 'secondary']),
+  activePath: PropTypes.string,
+  mobilePath: PropTypes.string,
+  navCollapsed: PropTypes.bool,
+  pinnedSecondaryNav: PropTypes.bool,
+  pinnedTertiaryNav: PropTypes.bool,
+  updateNavOnMenuToggleClick: PropTypes.func,
+  updateNavOnItemHover: PropTypes.func,
+  updateNavOnItemBlur: PropTypes.func,
+  updateNavOnItemClick: PropTypes.func,
+  updateNavOnPinSecondary: PropTypes.func,
+  updateNavOnPinTertiary: PropTypes.func,
+  updateNavOnMobileSelection: PropTypes.func,
+  clearMobileSelection: PropTypes.func,
+  setActivePath: PropTypes.func,
+  setMobilePath: PropTypes.func,
+  forceHideSecondaryMenu: PropTypes.func,
+  hoverDelay: PropTypes.number,
+  hideDelay: PropTypes.number
+};
+
+const getNextDepth = depth =>
+  (depth === 'primary' && 'secondary') ||
+  (depth === 'secondary' && 'tertiary') ||
+  'primary';
+
+const deepestOf = (pri, sec, ter) => (pri && sec && ter) || (pri && sec) || pri;
+
+const getItemProps = props => ({
+  ...selectKeys(props, Object.keys(itemObjectTypes)),
+  subItems:
+    props.children &&
+    React.Children.count(props.children) > 0 &&
+    React.Children.map(props.children, child => getItemProps(child.props))
+});
+
+const provideNavContext = withContext(navContextTypes, providerProps => {
+  // The item prop doesn't get included in context, but must be passed to the provider
+  // In order to properly include primaryItem and secondaryItem in context.
+  const { item, primaryItem, secondaryItem } = providerProps;
+  const nextDepth = getNextDepth(providerProps.depth);
+  return {
+    // Only the keys that should be in context are included,
+    // so it is safe to spread extra props into the provider component.
+    ...selectKeys(providerProps, Object.keys(navContextTypes)),
+    depth: nextDepth,
+    primaryItem: nextDepth === 'secondary' ? item : primaryItem,
+    secondaryItem: nextDepth === 'tertiary' ? item : secondaryItem
+    // We don't need a tertiaryItem in context (see VerticalNavigationItem.getContextNavItems)
+  };
+});
+const consumeNavContext = getContext(navContextTypes);
+
+const NavContextProvider = provideNavContext(props => (
+  <React.Fragment>{props.children}</React.Fragment>
+));
+
+// WARNING: HACK! HAAAACK HACK HACK HACK WARNING THIS IS A HACK.
+// We only use this to apply magic body classes when the component is used in uncontrolled mode. TODO it's a prop now
+// And only for consistency-- the better solution is to manage these classes yourself in the application.
+const getBodyContentElement = () => {
```

Comment:
yep - remember discussing this one w/ you a bit back. Thanks for calling it out in the comments - ideally PF Next does not do these sort of things on the `<body>` or `<html>` tags. React's context typically renders inside the top most element after `<body>` (typically the `<div>` beneath `<body>`) and the state tree/redux store/application begins there...  so for that reason, in the future I would propose we may consider a `Layout` component possibly (and use that to handle this sort of layout css state)? I think the way Material UI does this is very nice, in the way that it renders Themes and ThemeProviders at the top most level:
https://material-ui-next.com/customization/themes/

Any classes that render on the `<body>` tag are highly suspect to me. That being said...having `VertNav` as the top most component beneath Redux/Application seems to be OK for now and matches our current PF3 specs...

Reply:
Just to be clear, this is not the `<body>` element. This is the main content area for the application.

Reply:
ahh - apologies, was conflating a different thing i think ;). thanks!